src/User.php

```diff
@@ -168,9 +138,136 @@ public static function set_current_user( $user_id ) {
+		return $user_id;
+	}
+
+	/**
+	 * Register a new customer and log them in. Assumes the user is not already signed in.
+	 *
+	 * @param array $userdata The user data from Klarna.
+	 * @return int|WP_Error The new user's ID or WP_Error.
+	 */
+	public function register_new_user( $userdata ) {
+		$user_id = wp_insert_user( $userdata );
+		if ( is_wp_error( $user_id ) ) {
+			return new WP_Error( 'register_new_user', 'could not create user' );
+		}
+
+		do_action( 'woocommerce_created_customer', $user_id, $userdata, false );
+		return $user_id;
+	}
+
+	/**
+	 * Extract the user data from the ID token.
+	 *
+	 * @param string $id_token The ID token.
+	 * @param string $refresh_token The refresh token.
+	 * @return array An userdata array to be consumed by wp_insert_user.
+	 */
+	public function get_user_data( $id_token, $refresh_token ) {
+		$id_token = wp_parse_args(
+			$id_token,
+			array(
+				'locale'          => str_replace( '-', '_', get_locale() ),
+				'billing_address' => array(),
+			)
+		);
+
+		$userdata = array(
+			'role'        => 'customer',
+			'user_login'  => sanitize_user( $id_token['email'] ),
+			'user_pass'   => wp_generate_password(),
+			'user_email'  => sanitize_email( $id_token['email'] ),
+			'first_name'  => sanitize_text_field( $id_token['given_name'] ),
+			'last_name'   => sanitize_text_field( $id_token['family_name'] ),
+			'description' => __( 'Sign in with Klarna', 'siwk' ),
+			'locale'      => $id_token['locale'],
+		);
+
+		// Clean fields, and use default values to avoid undefined index.
+		$billing_address = array_map(
+			function ( $field ) {
+				if ( empty( $field ) ) {
+					return '';
+				}
+				return wc_clean( $field );
+			},
+			$id_token['billing_address']
+		);
+
+		$userdata['meta_input'] = array(
+			'billing_first_name'    => $userdata['first_name'],
+			'billing_last_name'     => $userdata['last_name'],
+			'billing_city'          => $billing_address['city'] ?? '',
+			'billing_state'         => $billing_address['region'] ?? '',
+			'billing_country'       => $billing_address['country'] ?? '',
+			'billing_postcode'      => $billing_address['postal_code'] ?? '',
+			'billing_address_1'     => $billing_address['street_address'] ?? '',
+			'billing_address_2'     => $billing_address['street_address_2'] ?? '',
+			'billing_phone'         => $id_token['phone'] ?? '',
+			'billing_email'         => $userdata['user_email'],
+			'shipping_first_name'   => $userdata['first_name'],
+			'shipping_last_name'    => $userdata['last_name'],
+			'shipping_city'         => $billing_address['city'] ?? '',
+			'shipping_country'      => $billing_address['country'] ?? '',
+			'shipping_state'        => $billing_address['region'] ?? '',
+			'shipping_postcode'     => $billing_address['postal_code'] ?? '',
+			'shipping_address_1'    => $billing_address['street_address'] ?? '',
+			'shipping_address_2'    => $billing_address['street_address_2'] ?? '',
+			'shipping_phone'        => $id_token['phone'] ?? '',
```

Comment:
~~Im wondering if we should set the defaults to empty strings or if we should only update the meta data if the token contained the info. Since with this it could be the case that a existing customer has a address field in WooCommerce already, but Klarna does not return it for some reason, causing us to remove it. Or is there a need to always default to a empty string?~~
Nevermind, i see the check for empty fields later. Missed that before writing this. My other question still stands though.

Also i noticed the shipping_ fields are set to the billing address from the token, does klarna not return any shipping data? And should we then override it?

Reply:
Yes, Klarna doesn't expose [any scopes](https://docs.klarna.com/conversion-boosters/sign-in-with-klarna/integrate-sign-in-with-klarna/web-sdk-integration/#scopes-and-claims) for the shipping address. I've updated the code to not overwrite if it exists.